### PR TITLE
Rich text boxes: make the toolbar appear over existing content #7063

### LINKED ISSUE #7063: Rich text boxes: make the toolbar appear over existing content
Currently, when the user clicks on a rich text form field, the formatting toolbar is inserted above the form field, shifting all elements below it. When the user clicks outside the form field after editing it, the toolbar disappears and elements below it shifts again. This jumping of page content gets in the way of a smooth user user experience. It can also result in 'mishit clicks' i.e. user tries to click a button but ends up clicking something else because content moves with the click.

Options:
1. make the toolbar stay fixed all the time
2. make the toolbar appear on top of the existing content above the form field (I believe this is the default mode but we changed it to the current behavior in a misguided attempt to improve user experience -- my bad 😢  )

Before clicking:
image

After clicking:
image

### PR DESCRIPTION
**Pull request by HirdayGupta**
Fixes #7063 

Scoured the app for instances of the tinyMCE editor and **changed the inline property to false** where required.

Unable to deploy this to a staging server due to Issue #7108. Will do so once that is fixed. 

NOTE: For some reason, on the Instructor interface, the following flow results in an error: `Comments -> Comments For Students -> Edit Comment -> Save` This results in a display of `errorPage.jsp.` This error does not take place when inline property is true. Console output for error is [here.](https://gist.github.com/HirdayGupta/1a12ee7dbccc9e1b612eeeac27fa7a87)

All other changes work correctly on local server.

### CODE REVIEW

File: src/main/webapp/js/contextualcomments.es6 [hunk 1]
```diff
@@ -18,7 +18,7 @@ function prepareComments() {
     if (typeof richTextEditorBuilder !== 'undefined') {
         /* eslint-disable camelcase */ // The property names are determined by external library (tinymce)
         richTextEditorBuilder.initEditor('#commenttext', {
-            inline: true,
+            inline: false,
```
Comment (damithc) [inline src/main/webapp/js/contextualcomments.es6]: Non-expert question: is this a case where it defaults to `false` if we omit it?

Comment (HirdayGupta) [inline src/main/webapp/js/contextualcomments.es6]: @damithc Apologies. Upon closer review of the [tinyMCE documentation](https://www.tinymce.com/docs/configure/editor-appearance/#inline) and trying it out on the local server, you are correct. Omitting this parameter would allow the editor to default to `false.` Would you recommend new commits that do so?

Comment (damithc) [inline src/main/webapp/js/contextualcomments.es6]: > Would you recommend new commits that do so?

Yes. let's do that.

**Comment by damithc** [comment]
Just to be sure, can you include a screenshot/gif of the new behavior?

**Comment by HirdayGupta** [comment]
### Instructor Logged in:

**Creating and editing feedback session:**

fscreate

**Students --> (Select Course) --> Add Comment (for student/team):**

students- comments

**Comments --> Edit/Add comments in any of the sections:**

commentseditadd

### Student Logged In:

**Submitting response to Feedback Session:**

fsresponse

**Comment by HirdayGupta** [comment]
Changes are up on the staging server. Use this [join link](https://teammates-hirday.appspot.com/page/instructorCourseJoin?key=9C5CB6FEBA094877F6FC870EB6ACC78560D86F34188DF145360844662A06026F9C5CB6FEBA094877F6FC870EB6ACC785C91E6BB080F5B8351E2E5BF9D25AF0A59BF5012BFE0FD52B4541EE6A815FC142&instructorinstitution=TEAMMATES) to join as an instructor and see the changes.

**Comment by thenaesh** [comment]
@HirdayGupta It looks like you need to use GodMode to update the associated HTML in the tests. The entire toolbar is now getting chucked into the HTML the moment "edit" is pressed, whether or not the actual text field is in focus.

**Comment by HirdayGupta** [comment]
@thenaesh on it. Will send in new commits with updated expected HTML using god mode soon. 

Meanwhile, can someone help with this:

> NOTE: For some reason, on the Instructor interface, the following flow results in an error: Comments -> Comments For Students -> Edit Comment -> Save This results in a display of errorPage.jsp. This error does not take place when inline property is true. Console output for error is [here](https://gist.github.com/HirdayGupta/1a12ee7dbccc9e1b612eeeac27fa7a87).

**Comment by HirdayGupta** [comment]
@thenaesh I also need some assistance running the test. I downloaded the older version of firefox (46.0) and attempted to use the custom firefox path in test.properties. However TestNG says it was unable to find the executable for the path I specified. 

I downloaded the `firefox-46.0.mac-x86_64.sdk.tar.bz2` file from [here](https://ftp.mozilla.org/pub/firefox/releases/46.0/) and tried adding both the `firefox.app` and the `/bin` folders as my custom paths for the executables. Where am I going wrong?

**Comment by whipermr5** [comment]
@HirdayGupta That's the Firefox SDK... Download the [release `.dmg`](https://ftp.mozilla.org/pub/firefox/releases/46.0/mac/en-US/Firefox%2046.0.dmg) instead and install it normally, then the path can be left blank.

**Comment by HirdayGupta** [comment]
Thanks! @whipermr5. Now when I run the tests, I get the following error message:

```JAVA
java.lang.AssertionError: Please append a unique id (e.g your name) to each of the default account in test.properties in order to use God mode, e.g change alice.tmms to alice.tmms.<yourName>, charlie.tmms to charlie.tmms.<yourName>, etc.
	at org.testng.AssertJUnit.fail(AssertJUnit.java:59)
	at teammates.test.driver.TestProperties.verifyReadyForGodMode(TestProperties.java:142)
	at teammates.test.pageobjects.AppPage.regenerateHtmlFile(AppPage.java:886)
	at teammates.test.pageobjects.AppPage.testAndRunGodMode(AppPage.java:878)
	at teammates.test.pageobjects.AppPage.verifyHtmlPart(AppPage.java:861)
	at teammates.test.pageobjects.AppPage.verifyHtmlMainContent(AppPage.java:903)
	at teammates.test.cases.browsertests.InstructorFeedbackPageUiTest.testPublishAction(InstructorFeedbackPageUiTest.java:600)
	at teammates.test.cases.browsertests.InstructorFeedbackPageUiTest.testRemindPublishActions(InstructorFeedbackPageUiTest.java:107)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:659)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:845)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1153)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at org.testng.TestRunner.privateRun(TestRunner.java:771)
	at org.testng.TestRunner.run(TestRunner.java:621)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:357)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:352)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:310)
	at org.testng.SuiteRunner.run(SuiteRunner.java:259)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1199)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1124)
	at org.testng.TestNG.run(TestNG.java:1032)
	at org.testng.remote.AbstractRemoteTestNG.run(AbstractRemoteTestNG.java:132)
	at org.testng.remote.RemoteTestNG.initAndRun(RemoteTestNG.java:230)
	at org.testng.remote.RemoteTestNG.main(RemoteTestNG.java:76)

```

However, in my test.properties file, I have made the changes being requested by the error trace. I have changed `alice.tmms` to `alice.tmms.HirdayGupta` but it still displays the same error.

**Comment by VamsiSangam** [comment]
Hi @HirdayGupta ! GodMode is meant to be run only on your local machine, not on Travis. Did you forget to turn off your GodMode? [This](https://github.com/TEAMMATES/teammates/issues/7186#issuecomment-296435908) should help you. :smile:

**Comment by HirdayGupta** [comment]
@VamsiSangam @whipermr5 I am aware that god mode is meant to be run locally 😄. That was the error on the local machine itself. I have not sent Travis any commits with God Mode on.

**Comment by VamsiSangam** [comment]
Oh sorry :sweat_smile: ! Well, I remember something similar happening to me before. Did you do the required changes in `test.template.properties` file too? I think I remember that changing it only in `test.properties` did not work for me too. Make the changes in `test.template.properties` too and clean your project.

**Comment by HirdayGupta** [comment]
@VamsiSangam thanks for the quick response! Unfortunately, even doing that doesn't seem to be working. I am still getting the same error on my local dev machine. Here is my[ `test.properties`](https://gist.github.com/HirdayGupta/24e616a678d568963f4e843553656757) file. Any idea what might be going wrong?

**Comment by VamsiSangam** [comment]
You must append the unique id to the [admin account](https://gist.github.com/HirdayGupta/24e616a678d568963f4e843553656757#file-test-properties-hirdaygupta-teammates-L58) too, and I suggest you append it to the [unregistered google account](https://gist.github.com/HirdayGupta/24e616a678d568963f4e843553656757#file-test-properties-hirdaygupta-teammates-L73) too. :smile: 

Actually because your admin account was not appended with the unique id, this is the method which gets called and I have commented on the line which leads to the error -

``` java
private static boolean areTestAccountsReadyForGodMode() {
    if (!TEST_STUDENT1_ACCOUNT.startsWith("alice.tmms.")) {
        return false;
    }
    String uniqueId = TEST_STUDENT1_ACCOUNT.substring("alice.tmms.".length());
    if (uniqueId.isEmpty()) {
        return false;
    }

    boolean isSecondStudentAccountReady = ("charlie.tmms." + uniqueId).equals(TEST_STUDENT2_ACCOUNT);
    boolean isInstructorAccountReady = ("teammates.coord." + uniqueId).equals(TEST_INSTRUCTOR_ACCOUNT);
    boolean isAdminAccountReady = ("yourGoogleId." + uniqueId).equals(TEST_ADMIN_ACCOUNT); // this fails
    return isSecondStudentAccountReady && isInstructorAccountReady && isAdminAccountReady;
}
```

Let me know if this works.

**Comment by HirdayGupta** [comment]
@VamsiSangam thank you for the assistance! 😄 That has solved my problem.

**Comment by HirdayGupta** [comment]
@damithc @whipermr5 new commits update the expected HTML files wherever necessary. I have split these updates into two commits - **requesting you to doubly review the changes in the second one (https://github.com/TEAMMATES/teammates/pull/7117/commits/d989b38600257bb44a39ec020d1b544ef526726c) as some of the changes by god-mode seemed unnecessary.**

Additionally, there are still some failing test cases, the causes of which I am unsure of. 

- The error thrown by `teammates.test.cases.browsertests.InstructorCommentsPageUiTest` I believe is due to the following that I have previously brought to notice and requested assistance with:
> NOTE: For some reason, on the Instructor interface, the following flow results in an error: Comments -> Comments For Students -> Edit Comment -> Save This results in a display of errorPage.jsp. This error does not take place when inline property is true. Console output for error is [here](https://gist.github.com/HirdayGupta/1a12ee7dbccc9e1b612eeeac27fa7a87).

**Comment by VamsiSangam** [comment]
Hi @HirdayGupta ! I think I found the source of your error. To simulate your work on my PC, I commented the line `inline: true` for `instructorComments.js`. Then ran the application and navigated to `Comments` -> `Comments for students` -> `Edit this comment`. The editor was up without clicking on anything, so it does look like you've solved the issue from the UI's point-of-view. When when I added some text to the comment and hit `Submit`, this was what I got in my debugger tools -

with-inline

If you notice, `commenttext` is empty and the TinyMCE editor is sending the text via another parameter named `commentText1`. You are getting an error because in `InstructorStudentCommentEditAction.java`, it looks for a parameter with name `commenttext`, the line is -

``` java
String commentTextString = getRequestParamValue(Const.ParamsNames.COMMENT_TEXT);
```

`COMMENT_TEXT` has the value `commenttext`. To double check my theory, I removed the comment in `instructorComments.js` and tried the same thing again (this time we have the clicking on text area issue, obviously), this was what I got in my debugger tools -

with-inline-2

## In a nutshell

I believe removing `inline: true` is the solution to the issue. But when you do that TinyMCE editor is sending the comment text over a parameter named `commentText1` rather than `commenttext`.

Why is it sending over `commentText1` suddenly? How to change it to `commenttext`? I'll leave those to you :stuck_out_tongue_closed_eyes:

**Comment by HirdayGupta** [comment]
Great catch @VamsiSangam! Will try figuring those last two questions out 😆.

**Comment by thenaesh** [comment]
Closing due to inactivity.
